### ISSUE TITLE
Task stuck on __wait_on_freeing_inode / bch2_inode_hash_find

### ISSUE THREAD

**Issue by g2p** [koverstreet/bcachefs#755]
Here is the backtrace from /proc/$pid/stack:
```
__wait_on_freeing_inode (fs/inode.c:2393) 
bch2_inode_hash_find (fs/bcachefs/fs.c:211) 
bch2_lookup_trans (fs/bcachefs/fs.c:512) 
bch2_lookup (fs/bcachefs/fs.c:556 (discriminator 13)) 
__lookup_slow (fs/namei.c:1732) 
walk_component (./include/linux/fs.h:830 fs/namei.c:1750 fs/namei.c:2053) 
path_lookupat (fs/namei.c:2556 fs/namei.c:2580) 
filename_lookup (fs/namei.c:2610 (discriminator 1)) 
vfs_statx (fs/stat.c:313) 
vfs_fstatat (fs/stat.c:342) 
__do_sys_newfstatat (fs/stat.c:506) 
__x64_sys_newfstatat (fs/stat.c:499) 
x64_sys_call (arch/x86/entry/syscall_64.c:36) 
do_syscall_64 (arch/x86/entry/common.c:52 (discriminator 1) arch/x86/entry/common.c:83 (discriminator 1)) 
entry_SYSCALL_64_after_hwframe (arch/x86/entry/entry_64.S:130) 
```

This is with bcachefs-testing (022d50aedff3b9961f6b8df579373a2b966f0f2a)

Let me know if there's anything I can do to help debug this

**Comment by g2p** [comment]
I can get the offending ino with drgn, but it doesn't appear in btrees/inodes/keys, probably because it's already deleted and the wakeup was lost.

**Comment by g2p** [comment]
On c4262a514ab6ae7397be3f7a789edda26ff38410, with the same command that I used to reproduce the initial issue (`find /mnt/bcachefs -size 42c`), things get stuck in a different fashion.

The first find run stats 600k files in 6 seconds, the second causes kernel warnings about a soft lockup and takes 35s, the third takes about 19m before registering Ctrl-C.

<details><summary>dmesg</summary>

<https://sebsauvage.net/paste/?f709b2c1a8d2e88c#C4GUJhQedXv+QlbJqTxgoqopy87TfOvCNlXvoVrIMFc=>
</details>

Decoded stack:
```
15:59:02.911783:  <TASK>
15:59:02.911794: ? asm_sysvec_apic_timer_interrupt (./arch/x86/include/asm/idtentry.h:702) 
15:59:02.911807: ? rhashtable_insert_slow (./include/linux/err.h:81 (discriminator 1) lib/rhashtable.c:628 (discriminator 1) lib/rhashtable.c:644 (discriminator 1)) 
15:59:02.911816: ? rhashtable_insert_slow (lib/rhashtable.c:513 (discriminator 3) lib/rhashtable.c:619 (discriminator 3) lib/rhashtable.c:644 (discriminator 3)) 
15:59:02.911824: bch2_inode_hash_insert (./include/linux/rhashtable.h:738 ./include/linux/rhashtable.h:914 fs/bcachefs/fs.c:316) 
15:59:02.911832: bch2_inode_hash_init_insert (fs/bcachefs/fs.c:420) 
15:59:02.911843: bch2_lookup_trans (./include/linux/err.h:116 (discriminator 1) fs/bcachefs/fs.c:589 (discriminator 1)) 
15:59:02.911852: bch2_lookup (fs/bcachefs/fs.c:626 (discriminator 13)) 
15:59:02.911860: ? bch2_lookup (fs/bcachefs/fs.c:626 (discriminator 13)) 
15:59:02.911869: __lookup_slow (fs/namei.c:1732) 
15:59:02.911877: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.911887: walk_component (./include/linux/fs.h:830 fs/namei.c:1750 fs/namei.c:2053) 
15:59:02.911896: path_lookupat (fs/namei.c:2556 fs/namei.c:2580) 
15:59:02.911904: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.911913: filename_lookup (fs/namei.c:2610 (discriminator 1)) 
15:59:02.911921: vfs_statx (fs/stat.c:313) 
15:59:02.911929: vfs_fstatat (fs/stat.c:342) 
15:59:02.911940: __do_sys_newfstatat (fs/stat.c:506) 
15:59:02.911948: __x64_sys_newfstatat (fs/stat.c:499) 
15:59:02.911957: x64_sys_call (arch/x86/entry/syscall_64.c:36) 
15:59:02.911968: do_syscall_64 (arch/x86/entry/common.c:52 (discriminator 1) arch/x86/entry/common.c:83 (discriminator 1)) 
15:59:02.911978: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.911987: ? do_syscall_64 (arch/x86/entry/common.c:102) 
15:59:02.911995: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912003: ? syscall_exit_to_user_mode (./arch/x86/include/asm/entry-common.h:57 (discriminator 1) ./arch/x86/include/asm/entry-common.h:65 (discriminator 1) ./include/linux/entry-common.h:330 (discriminator 1) kernel/entry/common.c:207 (discriminator 1) kernel/entry/common.c:218 (discriminator 1)) 
15:59:02.912012: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912022: ? do_syscall_64 (arch/x86/entry/common.c:102) 
15:59:02.912033: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912041: ? atime_needs_update (./include/linux/fs.h:2336 (discriminator 2) fs/inode.c:2079 (discriminator 2)) 
15:59:02.912051: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912060: ? touch_atime (fs/inode.c:2109 (discriminator 1)) 
15:59:02.912068: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912079: ? iterate_dir (./include/linux/fs.h:831 fs/readdir.c:113) 
15:59:02.912087: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912095: ? __f_unlock_pos (fs/file.c:1195) 
15:59:02.912104: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912112: ? __x64_sys_getdents64 (./include/linux/file.h:68 ./include/linux/file.h:85 fs/readdir.c:420 fs/readdir.c:392 fs/readdir.c:392) 
15:59:02.912120: ? __x64_sys_getdents64 (fs/readdir.c:350) 
15:59:02.912129: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912142: ? syscall_exit_to_user_mode (./arch/x86/include/asm/entry-common.h:57 (discriminator 1) ./arch/x86/include/asm/entry-common.h:65 (discriminator 1) ./include/linux/entry-common.h:330 (discriminator 1) kernel/entry/common.c:207 (discriminator 1) kernel/entry/common.c:218 (discriminator 1)) 
15:59:02.912150: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912161: ? do_syscall_64 (arch/x86/entry/common.c:102) 
15:59:02.912169: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912177: ? __x64_sys_fcntl (fs/fcntl.c:586 fs/fcntl.c:571 fs/fcntl.c:571) 
15:59:02.912186: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912194: ? syscall_exit_to_user_mode (./arch/x86/include/asm/entry-common.h:57 (discriminator 1) ./arch/x86/include/asm/entry-common.h:65 (discriminator 1) ./include/linux/entry-common.h:330 (discriminator 1) kernel/entry/common.c:207 (discriminator 1) kernel/entry/common.c:218 (discriminator 1)) 
15:59:02.912203: ? srso_return_thunk (arch/x86/lib/retpoline.S:224) 
15:59:02.912213: ? do_syscall_64 (arch/x86/entry/common.c:102) 
15:59:02.912222: entry_SYSCALL_64_after_hwframe (arch/x86/entry/entry_64.S:130) 
15:59:02.912230: RIP: 0033:0x7f63b3f173be
15:59:02.912241: Code: 0f 1f 40 00 48 8b 15 59 ba 0e 00 f7 d8 64 89 02 b8 ff ff ff ff c3 66 0f 1f 44 00 00 f3 0f 1e fa 41 89 ca b8 06 01 00 00 0f 05 <3d> 00 f0 ff ff 77 0b 31 c0 c3 0f 1f 84 00 00 00 00 00 48 8b 15 21
All code
========
   0:	0f 1f 40 00          	nopl   0x0(%rax)
   4:	48 8b 15 59 ba 0e 00 	mov    0xeba59(%rip),%rdx        # 0xeba64
   b:	f7 d8                	neg    %eax
   d:	64 89 02             	mov    %eax,%fs:(%rdx)
  10:	b8 ff ff ff ff       	mov    $0xffffffff,%eax
  15:	c3                   	ret
  16:	66 0f 1f 44 00 00    	nopw   0x0(%rax,%rax,1)
  1c:	f3 0f 1e fa          	endbr64
  20:	41 89 ca             	mov    %ecx,%r10d
  23:	b8 06 01 00 00       	mov    $0x106,%eax
  28:	0f 05                	syscall
  2a:*	3d 00 f0 ff ff       	cmp    $0xfffff000,%eax		<-- trapping instruction
  2f:	77 0b                	ja     0x3c
  31:	31 c0                	xor    %eax,%eax
  33:	c3                   	ret
  34:	0f 1f 84 00 00 00 00 	nopl   0x0(%rax,%rax,1)
  3b:	00 
  3c:	48                   	rex.W
  3d:	8b                   	.byte 0x8b
  3e:	15                   	.byte 0x15
  3f:	21                   	.byte 0x21

Code starting with the faulting instruction
===========================================
   0:	3d 00 f0 ff ff       	cmp    $0xfffff000,%eax
   5:	77 0b                	ja     0x12
   7:	31 c0                	xor    %eax,%eax
   9:	c3                   	ret
   a:	0f 1f 84 00 00 00 00 	nopl   0x0(%rax,%rax,1)
  11:	00 
  12:	48                   	rex.W
  13:	8b                   	.byte 0x8b
  14:	15                   	.byte 0x15
  15:	21                   	.byte 0x21
15:59:02.912250: RSP: 002b:00007ffd00d235b8 EFLAGS: 00000202 ORIG_RAX: 0000000000000106
15:59:02.912258: RAX: ffffffffffffffda RBX: 00007ffd00d236c0 RCX: 00007f63b3f173be
15:59:02.912267: RDX: 00007ffd00d236c0 RSI: 000055fdecf74f60 RDI: 0000000000000008
15:59:02.912276: RBP: 00007ffd00d235e0 R08: 0000000000000004 R09: 0000000000000007
15:59:02.912286: R10: 0000000000000100 R11: 0000000000000202 R12: 000055fdecf74f60
15:59:02.912297: R13: 000055fdecc20a20 R14: 0000000000000000 R15: 0000000000000004
15:59:02.912305:  </TASK>
```

**Comment by g2p** [comment]
Seems to be fixed (tested bcachefs-for-upstream 0f25eb4b60771f08fbcca878a8f7f88086d0c885 which has 6b63a948a73ba3df0fb3ab0c44807df344bc5bbf)
